### PR TITLE
Allow automatic actions to save recompution

### PR DESCRIPTION
**Pull request by jenf**
Here's an alternative approach to #3223

@ventusignis for review as well to see if it'd work.

I've tested server based games, migrations (although there's a flaw if the culprit is the action inside auto actions)

In theory this is all working now

### CODE REVIEW

**Comment by jenf** [comment]
The stored data would look like
` {
      "type": "lay_tile",
      "entity": "ICR",
      "entity_type": "corporation",
      "id": 647,
      "created_at": 1611507962,
      "derived": [
        {
          "type": "pass",
          "entity": "ICR",
          "entity_type": "corporation",
          "created_at": 1611507962
        }
      ],
      "hex": "H12",
      "tile": "9-12",
      "rotation": 0
    },`

File: lib/engine/game/base.rb [hunk 1]
```diff
@@ -644,13 +675,11 @@ def process_action(action)
         end
 
         @last_processed_action = action.id
-        self
```
Comment (tobymao) [inline lib/engine/game/base.rb:647]: why are these removed?

File: lib/engine/step/g_1867/token.rb [hunk 2]
```diff
@@ -7,6 +7,18 @@ module Engine
...
     module G1867
       class Token < Token
+        def derived_actions(entity)
+          Engine::Action::Pass.new(entity) unless can_really_place_token?(entity)
+        end
+
+        def can_really_place_token?(entity)
+          # Cheaper to do the graph first, then check affordability
+          current_entity == entity &&
+            (tokens = available_tokens(entity)).any? &&
```
Comment (tobymao) [inline lib/engine/step/g_1867/token.rb:17]: !empty

Comment (tobymao) [reply lib/engine/step/g_1867/token.rb:17]: unless tokens contains nils

File: lib/engine/round/base.rb [hunk 3]
```diff
@@ -130,6 +130,10 @@ def active_step(entity = nil)
         @active_step ||= @steps.find { |step| step.active? && step.blocking? }
       end
 
+      def derived_actions
```
Comment (tobymao) [inline lib/engine/round/base.rb:133]: do we need derived actions? can't we just do @game.add_x_action(action)

def add_x_action(action)
  raise if @loading

File: lib/engine/game/base.rb [hunk 4]
```diff
@@ -597,16 +598,46 @@ def initialize_actions(actions)
...
 
+        # Process the action we came here to do first
+        process_single_action(action)
+
+        # Process any derived actions
+        # derived actions cannot be undos so we don't need to worry about state
+        action.derived.each { |a| process_single_action(a) }
+        any_derived = !action.derived.empty?
+
+        unless @loading_internal
```
Comment (tobymao) [inline lib/engine/game/base.rb:621]: is this whole section necessary? 

can add_action(x) just mutate the state of the actions?

and then we just do 

while (action = action.derived_actions.pop)

Comment (tobymao) [reply lib/engine/game/base.rb:621]: we can rename to auto action

File: lib/engine/game/base.rb (hunk 1 above)
Comment (jenf) [reply lib/engine/game/base.rb:647]: they were originally used for the undo functionality, since this isn't the process action it makes no sense

File: lib/engine/game/base.rb (hunk 4 above)
Comment (jenf) [reply lib/engine/game/base.rb:621]: not really, if you have chained auto actions this wouldn't work since the step can change

Comment (tobymao) [reply lib/engine/game/base.rb:621]: i think it should work if it's done in a while looop?

File: lib/engine/round/base.rb (hunk 3 above)
Comment (jenf) [reply lib/engine/round/base.rb]: as discussed, auto_actions need to be able to be doable on a transition between steps so this is probably the cleanest way of doing it.

File: lib/engine/game/base.rb [hunk 5]
```diff
@@ -704,9 +734,7 @@ def next_turn!
       end
 
       def action_from_h(h)
-        Object
-          .const_get("Engine::Action::#{Action::Base.type(h['type'])}")
-          .from_h(h, self)
+        Engine::Action::Base.action_from_h(h, self)
```
Comment (ventusignis) [inline lib/engine/game/base.rb:737]: Should we remove this method and replace its calls with just creating the action where it's used?

Comment (jenf) [reply lib/engine/game/base.rb:737]: good call

File: lib/engine/step/g_1867/token.rb [hunk 6]
```diff
@@ -7,6 +7,18 @@ module Engine
   module Step
     module G1867
       class Token < Token
+        def auto_action(entity)
+          Engine::Action::Pass.new(entity) unless can_really_place_token?(entity)
```
Comment (jenf) [inline lib/engine/step/g_1867/token.rb:11]: One slight flaw with how this is done is it'll log a pass rather than skipping

A solution would be to store if the Pass was a "skip" and tweak the debug appropriately

File: lib/engine/game/base.rb (hunk 4 above)
Comment (jenf) [reply lib/engine/game/base.rb:621]: as mentioned in chat pop mutates the list which is needed outside to actually store the auto_actions.

I'll try and see if there's any tidying up but my initial attempts don't seem to be going anywhere

**Comment by jenf** [comment]
I think this is as close as I can get it right now

File: lib/engine/round/base.rb [hunk 7]
```diff
@@ -130,6 +130,10 @@ def active_step(entity = nil)
         @active_step ||= @steps.find { |step| step.active? && step.blocking? }
       end
 
+      def auto_action
+        active_step(nil)&.auto_action(current_entity)
```
Comment (tobymao) [inline lib/engine/round/base.rb:134]: you can just call active_step

File: lib/engine/step/base.rb [hunk 8]
```diff
@@ -32,6 +32,8 @@ def actions(_entity)
         []
       end
 
+      def auto_action(_entity); end
```
Comment (tobymao) [inline lib/engine/step/base.rb:35]: should this be named auto_actions

Comment (jenf) [reply lib/engine/step/base.rb:35]: yep

File: routes/game.rb [hunk 9]
```diff
@@ -83,7 +83,7 @@ class Api
 
                 r.params['user'] = user.id
 
-                engine = engine.process_action(r.params)
+                engine = engine.process_action(r.params, add_autoactions: false)
```
Comment (tobymao) [inline routes/game.rb:86]: why is this false? isn't this the only place you'd want it to be true? maybe i don't understand what this param is doing

Comment (jenf) [reply routes/game.rb:86]: The way it's designed is that only the client actually adds the autoactions and so the server won't then add them. This would cause inconsistency between the server and the client if we're mid-deploy.
I could allow the server to add them and hope the consistency remains the same, but I'm wary of there being races here
